### PR TITLE
feat(maven-use): improve selection when rewriting version numbers

### PR DESCRIPTION
**Pull request by bepri**
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you added an entry to the changelog (`docs/reference/changelog.rst`)?

-----

The logic for this is unfortunately quite messy since I'm dealing with potentially arbitrary strings, but I tried to ensure each code path was documented and had full test coverage. I'm open to adding more tests if some edge cases are thought of.

### CODE REVIEW

File: docs/reference/changelog.rst [hunk 1]
```diff
@@ -22,11 +22,24 @@ Changelog
...
+New features:
+
+- Previously, when the Maven Use plugin updated ``pom.xml`` for self-contained projects,
+  it wouldn't reliably find the correct dependency versions on the host. It could
+  unpredictably declare the wrong package version, or select a vastly different version
+  despite a similar one being available.
+
+  With Craft Parts 2.20.1, the plugin now deterministically detects and matches the
+  dependency versions available on the host, aligning with how Maven normally behaves.
+  This change makes the plugin a drop-in replacement of Maven, for private networks.
```
Comment (medubelko) [inline docs/reference/changelog.rst:34]:
```suggestion
  This change makes the plugin a drop-in replacement for Maven in private networks.
```

My bad, I was too loose with the prepositions.

File: docs/reference/changelog.rst [hunk 2]
```diff
@@ -22,11 +22,24 @@ Changelog
...
+
 Bug fixes:
 
 - `#1007 <https://github.com/canonical/craft-parts/issues/1007>`_ When wild cards
   were used in an ``organize`` source path, an error would occur if files mapped to
   themselves. These cases are now ignored.
+- When rebuilding a part with the ``self-contained`` build attribute, the maven-use
```
Comment (medubelko) [inline docs/reference/changelog.rst:41]:
```suggestion
- When rebuilding a part with the ``self-contained`` build attribute, the Maven Use
```

File: docs/reference/changelog.rst [hunk 3]
```diff
@@ -22,11 +22,24 @@ Changelog
...
+  plugin now guarantees to select the same dependency version.
```
Comment (medubelko) [inline docs/reference/changelog.rst:42]: I honestly think we could skip this, since it's covered in the new features section.

**Review by medubelko — APPROVED** [review]
Docs look good! Do you want to sit on it for a bit before considering the changelog final?

**Comment by bepri** [comment]
@medubelko I think as long as the engineer reviews agree with the general accuracy. I'm happy with the content myself.

File: tests/unit/utils/test_maven_utils.py [hunk 4]
```diff
@@ -346,6 +425,207 @@ def to_pom(self, repository: Path) -> Path:
...
+            "mint",
+            id="two-fallbacks",
+        ),
+    ],
+)
+def test_versions_max(versions: _Versions, expected: str) -> None:
+    assert versions.max() == expected
+
+
+@pytest.mark.parametrize(
```
Comment (tigarmo) [inline tests/unit/utils/test_maven_utils.py:477]: nice test cases 👍🏾

**Review by tigarmo — APPROVED** [review]
looks nice, thanks!

**Review by lengau — APPROVED** [review]
